lineax/_operator.py

```diff
@@ -906,6 +906,78 @@ def out_structure(self):
         return self.operator.out_structure()
 
 
+class WoodburyLinearOperator(AbstractLinearOperator, strict=True):
+    """As [`lineax.MatrixLinearOperator`][], but for specifically a matrix
+    with A + U C V structure, such that the Woodbury identity can be used"""
+
+    A: AbstractLinearOperator
+    C: Inexact[Array, " k k"]
+    U: Inexact[Array, " n k"]
+    V: Inexact[Array, " k n"]
+    UCV: Array
+    tags: frozenset[object] = eqx.field(static=True)
+
+    def __init__(
+        self,
+        A: AbstractLinearOperator,
+        C: Inexact[Array, " k k"],
+        U: Inexact[Array, " n k"],
+        V: Inexact[Array, " k n"],
+        tags: Union[object, frozenset[object]] = (),
+    ):
+        """**Arguments:**
+
+        Matrix of form A + U C V, such that the inverse can be computed
+        using Woodbury matrix identity
+
+        - `A`: Linear operator, in/out shape (n,n)
+        - `C`: A rank-two JAX array. Shape (k,k)
+        - `U`: A rank-two JAX array. Shape (n,k)
+        - `V`: A rank-two JAX array. Shape (k,n)
+
+        """
+        self.A = A
+        self.C = inexact_asarray(C)
+        self.U = inexact_asarray(U)
+        self.V = inexact_asarray(V)
+        (N, M) = self.A.in_structure(), self.A.out_structure()
+        if N != M:
+            raise ValueError(f"expecting square operator for A, got {N} by {M}")
+        (K, L) = self.C.shape
+        if K != L:
+            raise ValueError(f"expecting square operator for C, got {K} by {L}")
+        N = N.shape[0]
+        if self.U.shape != (N, K):
+            raise ValueError("U does not have consistent shape with A and C")
+        if self.V.shape != (K, N):
+            raise ValueError("V does not have consistent shape with A and C")
+        self.UCV = self.U @ (self.C @ self.V)
```

Comment:
Storing things like this is probably a bug. If we modify this operator in any other way than through `__init__` (e.g. evolving it under gradient descent) then this invariant may cease to hold.

Better to compute it on-the-fly when required instead.

Reply:
(In addition this is quite space inefficient!)